File: experimental/tools/elastic-miter-generator/ndwirerer.cpp
#include "../experimental/tools/elastic-miter-generator/CreateStateWrapper.h"
#include "../experimental/tools/elastic-miter-generator/GetStates.h"
#include "dynamatic/InitAllDialects.h"
#include "mlir/IR/BuiltinOps.h"
#include "mlir/Parser/Parser.h"
#include "llvm/Support/InitLLVM.h"
#include <cstdlib>
#include <filesystem>
#include <fstream>
#include <iostream>
#include <string>

using namespace mlir;
using namespace llvm;

// TODO ...
const std::string OUT_DIR = "experimental/tools/elastic-miter-generator/out";
const std::string COMP_DIR = OUT_DIR + "/comp";
const std::string DOT = COMP_DIR + "/miter.dot";
const std::string REWRITES =
    "experimental/test/tools/elastic-miter-generator/rewrites";
const std::string MOD = "z";

void exitOnFail(int ret, const std::string &message) {
  if (ret != 0) {
    std::cerr << message << std::endl;
    exit(1);
  }
}

int runNuXmv(const std::string &cmd, const std::string &stdoutFile) {
  std::string command = "nuXmv -source " + cmd + " > " + stdoutFile;
  return system(command.c_str());
}

FailureOr<std::string> handshake2smv(const std::string &mlir,
                                     bool png = false) {
  std::string cmd = "bin/export-dot " + mlir + " --edge-style=spline > " + DOT;
  int ret = system(cmd.c_str());
  exitOnFail(ret, "Failed to convert to dot");

  if (png) {
    cmd = "dot -Tpng " + DOT + " -o " + COMP_DIR + "/visual.png";
    ret = system(cmd.c_str());
    exitOnFail(ret, "Failed to convert to PNG");
  }

  cmd = "python3 ../dot2smv/dot2smv " + DOT;
  ret = system(cmd.c_str());
  if (ret != 0) {
    llvm::errs() << "Failed to convert to SMV\n";
    return failure();
  }

  std::filesystem::rename(
      std::filesystem::path(COMP_DIR + "/model.smv"),
      std::filesystem::path(COMP_DIR + "/" + MOD + "_lhs.smv"));
  return COMP_DIR + "/" + MOD + "_lhs.smv";
}

int main(int argc, char **argv) {
  llvm::InitLLVM y(argc, argv);

  DialectRegistry registry;
  dynamatic::registerAllDialects(registry);
  MLIRContext context(registry);

  auto failOrDstSmv = handshake2smv(REWRITES + "/" + MOD + "_lhs.mlir", true);
  if (failed(failOrDstSmv))
    return 1;

  std::string dstSmv = failOrDstSmv.value();

  OwningOpRef<ModuleOp> modRef =
      parseSourceFile<ModuleOp>(REWRITES + "/" + MOD + "_lhs.mlir", &context);

  ModuleOp mod = modRef.release();

  // TODO use correct name
  // # TODO we should probably just pass the filename
  // Create state wrapper for infinite tokens
  auto failOrInfWrapper =
      dynamatic::experimental::createStateWrapper("z_lhs.smv", mod, 0, true);
  if (failed(failOrInfWrapper))
    return 1;

  std::string infWrapper = failOrInfWrapper.value();

  std::ofstream infFile(OUT_DIR + "/comp/main_inf.smv");
  infFile << infWrapper;
  infFile.close();

  // Run nuXmv for infinite tokens
  int ret = runNuXmv(COMP_DIR + "/prove_inf.cmd", OUT_DIR + "/inf_states.txt");
  exitOnFail(ret, "Failed to analyze reachable states with infinite tokens.");

  int n = 1;
  while (true) {
    llvm::outs() << "Checking " << n << " tokens.\n";

    // TODO use correct name
    auto failOrFinWrapper =
        dynamatic::experimental::createStateWrapper("z_lhs.smv", mod, n, false);
    if (failed(failOrFinWrapper))
      return 1;

    std::string finWrapper = failOrFinWrapper.value();

    std::ofstream finFile(OUT_DIR + "/comp/main_" + std::to_string(n) + ".smv");
    finFile << finWrapper;
    finFile.close();

    // TODO automatically create cmd file
    ret = runNuXmv(COMP_DIR + "/prove_" + std::to_string(n) + ".cmd",
                   OUT_DIR + "/" + std::to_string(n) + "_states.txt");
    exitOnFail(ret, "Failed to analyze reachable states with " +
                        std::to_string(n) + " tokens.");

    // Check state differences
    int nrOfDifferences = dynamatic::experimental::getStates(
        OUT_DIR + "/inf_states.txt",
        OUT_DIR + "/" + std::to_string(n) + "_states.txt");

    if (nrOfDifferences != 0) {
      n++;
    } else {
      std::cout << n << std::endl;
      break;
    }
  }
}

File: experimental/tools/elastic-miter-generator/ndwirerer.py
import os
import subprocess

from create_state_wrapper import create_state_wrapper
from get_states import get_states

# TODO clean this up, argparse
OUT_DIR = "experimental/tools/ndwirerer/out"

COMP_DIR = os.path.join(OUT_DIR, "comp")
DOT = os.path.join(COMP_DIR, "miter.dot")

REWRITES = "experimental/test/tools/elastic-miter-generator/rewrites"
MOD = "z"


def exit_on_fail(ret, message):
  if ret != 0:
    print(message)
    exit(1)

def run_nuXmv(cmd, stdout):
  cmd = ["nuXmv", "-source", cmd]
  with open(stdout, "w") as result_file:
    process = subprocess.run(cmd, stdout=result_file)
  return process.returncode


def handshake2smv(mlir, png=False):
  # Convert to DOT
  cmd = ["bin/export-dot", mlir, "--edge-style=spline"]
  with open(DOT, "w") as dot_file:
    process = subprocess.run(cmd, stdout=dot_file)
  exit_on_fail(process.returncode, "Failed to convert to dot")

  if png:
    # Convert to PNG
    cmd = ["dot", "-Tpng", DOT]
    with open(os.path.join(COMP_DIR, "visual.png"), "wb") as png_file:
      process = subprocess.run(cmd, stdout=png_file)
    exit_on_fail(process.returncode, "Failed to convert to PNG")

  # Convert to SMV
  cmd = ["python3", "../dot2smv/dot2smv", DOT]
  process = subprocess.run(cmd)
  exit_on_fail(process.returncode, "Failed to convert to SMV")

  # Move the SMV file
  src_smv = os.path.join(os.path.dirname(DOT), "model.smv")
  dst_smv = os.path.join(os.path.dirname(DOT), f"{MOD}_lhs.smv")
  os.rename(src_smv, dst_smv)

  return dst_smv

    

# TODO do this for LHS/RHS
def main():

  dst_smv = handshake2smv(f"{REWRITES}/{MOD}_lhs.mlir", True)

  # TODO we should probably just pass the filename 
  # Create state wrapper for infinite tokens
  inf_wrapper = create_state_wrapper(dst_smv, f"{REWRITES}/{MOD}_lhs.mlir", inf=True)
  with open("experimental/tools/ndwirerer/out/comp/main_inf.smv", "w") as f:
    f.write(inf_wrapper)
    
  # Run nuXmv for infinite tokens
  ret = run_nuXmv(os.path.join(COMP_DIR, "prove_inf.cmd"), os.path.join(OUT_DIR, f"inf_states.txt"))
  exit_on_fail(ret, "Failed to analyze reachable states with infinite tokens.")


  N = 1
  while True:
    print(f"Checking {N} tokens.")

    fin_wrapper = create_state_wrapper(dst_smv, f"{REWRITES}/{MOD}_lhs.mlir", N=N)
    with open(f"experimental/tools/ndwirerer/out/comp/main_{N}.smv", "w") as f:
      f.write(fin_wrapper)

    # TODO automatically create cmd file
    ret = run_nuXmv(os.path.join(COMP_DIR, f"prove_{N}.cmd"), os.path.join(OUT_DIR, f"{N}_states.txt"))
    exit_on_fail(ret, f"Failed to analyze reachable states with {N} tokens.")

    # Check state differences
    nr_of_differences = get_states(os.path.join(OUT_DIR, f"inf_states.txt"), os.path.join(OUT_DIR, f"{N}_states.txt"))

    if nr_of_differences != 0:
      N += 1
    else:
      print(N)
      break


if(__name__ == "__main__"):
  main()